File: src/sub/resolcovar.c
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h> 

#define IEQ 658
#define INSHOT 50
#define INLTOT 100
#define IST 650
#define ILIP (4 * IEQ + INSHOT + INLTOT + IST)
#define INVA (ILIP - 1)

extern bool single_turbo;
extern int isingle;
extern int neqs;
extern int nshot;
extern int nltot;
extern int iresolcalc;
extern int ifixsolution;
extern int nvar;
extern int ksta;

extern double g[];
extern double rht[];
extern double gcopy[60][INVA * 2];
extern double scale[];
extern double vdamp[][100];
extern int nplay[];
extern double rc[][4];
extern double covc[][4];
extern double r3[][3];
extern double spread;
extern double sv[];
extern double ale[];
extern double s[];

extern void luelmp(double a[], double b[], int n, double x[]);
extern void fixunt(double b[], int neqs_local, int nshot_local, int nltot_local, int ksta_local,
                   double scale_local[], double vdamp_local[][100], int itotmodels_local,
                   int inltot_local, int nplay1);
extern void spreadd(double r[][4], int n, double *spread_out);
extern void spreadb(double r[][4], int n, double *spread_out);

void resolcovar(double davari) {
    int mef = 4 * neqs + nshot - 4;
    int nef = 4 + nltot;
    if (isingle != 0) {
        nef = 4;
    }

    int n1 = 4 * neqs;
    int n2 = n1 + nshot;
    int n3 = n2 + nltot;

    double rdiag[100] = {0.0f};

    for (int row_index = 0; row_index < nef; ++row_index) {
        for (int col_index = 0; col_index < nvar; ++col_index) {
            rht[col_index] = gcopy[row_index][col_index];
        }

        int line_index = row_index;
        if (row_index >= 4) {
            line_index = row_index + mef;
        }

        luelmp(g, rht, nvar, rht);

        if (isingle == 0) {
            rdiag[row_index] = rht[line_index];
        } else {
            for (int column = 0; column < 4; ++column) {
                rc[line_index][column] = rht[column];
            }
        }

        luelmp(g, rht, nvar, rht);

        int block = (line_index + 1) % 4;
        if (block == 0) {
            block = 4;
        }

        double scale1 = scale[block - 1];
        if (line_index >= n1) {
            scale1 = scale[0];
        }
        if (line_index >= n2) {
            scale1 = scale[5];
        }
        if (line_index >= n3) {
            scale1 = scale[4];
        }

        for (int column = 0; column < nvar; ++column) {
            rht[column] = rht[column] * scale1;
        }

        fixunt(rht, neqs, nshot, nltot, ksta, scale, vdamp, 2, 100, nplay[0]);

        if (isingle != 0) {
            for (int column = 0; column < 4; ++column) {
                covc[line_index][column] = rht[column];
            }
        }

        s[row_index] = sqrtf(fabs(rht[line_index]));
    }

    if (isingle != 0) {
        double spread1 = 0.0f;
        double spread2 = 0.0f;

        spreadd(rc, 4, &spread1);
        spread = spread1;

        if (ifixsolution == 1) {
            for (int row = 0; row < 3; ++row) {
                for (int column = 0; column < 3; ++column) {
                    r3[row][column] = rc[row][column];
                }
            }
            double spread3 = 0.0f;
            spreadd((double (*)[4])r3, 3, &spread3);
            spread = spread3;
        }

        if (ifixsolution == 9) {
            double spread4 = 0.0f;
            spreadd((double (*)[4])rc, 1, &spread4);
            spread = spread4;
        }

        spreadb(rc, 4, &spread2);

        double size = 0.0f;
        for (int index = 0; index < 4; ++index) {
            size += covc[index][index];
        }

        if (!single_turbo) {
            fprintf(stdout, " D-Spread(R) = %6.3f   B-G Spread(R) = %6.3f       Size (C) = %6.3f\n",
                    spread, spread2, size);
        }

        double avresol = 0.0f;
        for (int index = 0; index < nef; ++index) {
            avresol += rc[index][index];
        }
        avresol /= (double)nef;
        (void)avresol;

        fprintf(stdout, "                        OT (sec)   X (km)    Y (km)    Z (km) \n");
        fprintf(stdout, " Sigma (CHD):         %10.4f%10.4f%10.4f%10.4f\n", s[0], s[1], s[2], s[3]);
        fprintf(stdout, " Resolution (CHD):    %10.4f%10.4f%10.4f%10.4f D-spread =%6.3f\n",
                rc[0][0], rc[1][1], rc[2][2], rc[3][3], spread);
        fprintf(stdout, " Data Variance      = %10.4f\n", davari);
        fprintf(stdout, " Singular values:     %10.4f%10.4f%10.4f%10.4f     ALE =%7.3f\n",
                sv[0], sv[1], sv[2], sv[3], ale[0]);
    }

    if (!single_turbo) {
        fprintf(stdout, "\n");
        fprintf(stdout, "Rdiag of selected model parameters:\n");
        for (int index = 0; index < nef; ++index) {
            fprintf(stdout, "%10.4f", rdiag[index]);
        }
        fprintf(stdout, "\n\n");
    }
}

File: src_fortran/sub/resolcovar.f
c
c
c
c
c
c
      subroutine RESOLCOVAR(davari)
c
c     calculate resolution and covariance matrices.
c
      implicit none
      real davari
      include '../inc/vel_com.inc'
c
      integer mef,nef,n1,n2,n3,k,i,l,j
      real scale1,spread1,spread2,size,avresol
      real Rdiag(inva)
c
      character*80 pcard
c
c---- resolution and covariance calculations
      if(.not.single_turbo)then
         write(16,'(/////)')
         write(16,40)
   40    format(' Resolution and covariance calculations:')
         if(isingle.ne.0)then
cc       write(16,*)'CHOLESKY-decomposition:'
         write(16,*)'    RESOLUTION-matrix                   ',
     &              '    COVARIANCE-matrix'
         write(16,*)
         endif
      endif
   25 mef=4*neqs+nshot-4
      nef=4+nltot
      if(isingle.ne.0) NEF=4     ! only first 4 rows and first 4 standard-devs.
      n1=4*neqs
      n2=n1+nshot
      n3=n2+nltot
      if(iresolcalc.eq.2.and.isingle.eq.1) 
     &write(80,*)'Covariance matrice'
      do 15 k=1,nef
      do 16 i=1,nvar
   16 rht(i)=gcopy(k,i)
      l=k
      if(k.gt.4) l=k+mef
      call LUELMP(g,rht,nvar,rht)
      if(isingle.eq.0)then
         if(.not.single_turbo)then
            write(16,17) l
 17         format(1x,'resolution row',i3)
            write(16,18)  (rht(j),j=1,nvar)
         endif
         Rdiag(k)=rht(l)
      else
         write(pcard,18) (rht(j),j=1,nvar)
         do j=1,4
            Rc(l,j)=rht(j)     !  resolution-matrix
         enddo
      endif
 18   format(8f10.4)
      call LUELMP(g,rht,nvar,rht)
c---- put covariance into proper unit
      j=mod(l,4)
      if(j.eq.0) j=4
      scale1=scale(j)
      if(l.gt.n1) scale1=scale(1)
      if(l.gt.n2) scale1=scale(6)
      if(l.gt.n3) scale1=scale(5)
      do 19 j=1,nvar
ccc   19 rht(j)=rht(j)*davari*scale1
   19 rht(j)=rht(j)*scale1  ! <== UNIT covariance matrix !!!
       call FIXUNT(rht,neqs,nshot,nltot,ksta,scale,
     &             vdamp,itotmodels,inltot,nplay(1))
      if(isingle.eq.0)then
         if(.not.single_turbo)then
            write(16,20) l
 20         format(1x,'covariance row',i3)
            write(16,18)  (rht(j),j=1,nvar)
         endif
      else
         write(pcard(41:80),18) (rht(j),j=1,nvar)
         if(.not.single_turbo) write(16,*) pcard
         do j=1,4
            COVc(l,j)=rht(j)
         enddo
      endif
c      s(k)=sqrt(abs(rht(l)))  ! compiler produces warning with range checking
      s(k)=abs(rht(l))
      s(k)=sqrt(s(k))
 15   continue
c
      if(isingle.ne.0)then
         if(.not.single_turbo) write(16,*)
         call SPREADd(Rc,4,spread1)
         spread=spread1
         if(ifixsolution.eq.1)then
            do j=1,3
               do k=1,3
                  R3(j,k)=Rc(j,k)
               enddo
            enddo
            call SPREADd(R3,3,spread)
         endif
         if(ifixsolution.eq.9)then
            call SPREADd(Rc(1,1),1,spread)
         endif
         call SPREADb(Rc,4,spread2)
         size=0.0
         do j=1,4
            size=size+COVc(j,j)
         enddo
         if(.not.single_turbo)then
            write(16,'(1x,''D-Spread(R) = '',f6.3,
     &                    ''   B-G Spread(R) = '',f6.3,
     &                 ''       Size (C) = '',f6.3
     &             )') spread,spread2, size
         endif
      endif
c
      if(.not.single_turbo)then
         write(16,*)
         write(16,21)
 21      format(' standard deviation of selected model parameters:')
      endif
      if(isingle.ne.0)then
        avresol=0.0
        do j=1,nef
           avresol=avresol+Rc(j,j)
        enddo
        avresol=avresol/float(nef)
        i f (.not.single_turbo) t h e n
        write(16,*)
        write(16,*)'   OT (sec)   X (km)    Y (km)    Z (km) '
        e n d i f
        write(6,'(23x,''  OT (sec)   X (km)    Y (km)    Z (km) '')')
c        write(6,'(1x,''CHOLESKY  (othet='',f5.3,'') :'')') othet
        write(6,'('' Sigma (CHD):         '',4f10.4)') (s(j),j=1,nef)
        write(6,'('' Resolution (CHD):    '',4f10.4,x,
     &            ''D-spread ='',f6.3)') (Rc(j,j),j=1,nef) , spread
        write(6,'('' Data Variance      = '',f10.4)') davari
ccc        call SPREADd(Rs,4,spread3)
ccc        do j=1,4
ccc           COVs(j)=SQRT(COVs(j))  !  <-- standard deviation
ccc        enddo
        write(6,'('' Singular values:     '',4f10.4,5x,''ALE ='',f7.3)')
     &               (SV(j),j=1,nef) , ale(1)
ccc        write(6,'('' Sigma (SVD):         '',4f10.4)') (COVs(j),j=1,nef)
ccc        write(6,'('' Resolution (SVD):    '',4f10.4,3x,
ccc     &            ''D-spread ='',f6.3)') (Rs(j,j),j=1,nef) , spread3
      endif
      if(.not.single_turbo) write(16,22) (s(j),j=1,nef)
 22   format(8f10.4)
      if(.not.single_turbo)then
         write(16,*)
         write(16,*)'Rdiag of selected model parameters:'
         write(16,22) (Rdiag(j),j=1,nef)
         write(16,*)
         write(16,*)
      endif
c
      return
      end ! of subr. resolcovar
